File: app/src/main/java/net/veldor/flibustaloader/parsers/SequencesParser.java
package net.veldor.flibustaloader.parsers;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.selections.FoundedSequence;

import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.ArrayList;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;

class SequencesParser {
    public static ArrayList parse(NodeList entries, XPath xPath) throws XPathExpressionException {
        ArrayList<FoundedSequence> result = new ArrayList<>();
        Node entry;
        FoundedSequence sequence;
        int counter = 0;
        int entriesLength = entries.getLength();
        int handledEntryCounter = 0;
        while ((entry = entries.item(counter)) != null) {
            ++handledEntryCounter;
            App.getInstance().mLoadAllStatus.postValue("Обрабатываю серию " + handledEntryCounter + " из " + entriesLength);
            sequence = new FoundedSequence();
            sequence.title = ((Node) xPath.evaluate("./title", entry, XPathConstants.NODE)).getTextContent();
            sequence.content = ((Node) xPath.evaluate("./content", entry, XPathConstants.NODE)).getTextContent();
            sequence.link = ((Node) xPath.evaluate("./link", entry, XPathConstants.NODE)).getAttributes().getNamedItem("href").getTextContent();
            result.add(sequence);
            counter++;
        }
        return result;
    }
}

File: app/src/main/java/net/veldor/flibustaloader/parsers/SequencesParser.kt
package net.veldor.flibustaloader.parsers

import kotlin.Throws
import net.veldor.flibustaloader.App
import net.veldor.flibustaloader.selections.FoundedSequence
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import java.util.ArrayList
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathExpressionException

internal object SequencesParser {
    @Throws(XPathExpressionException::class)
    fun parse(entries: NodeList, xPath: XPath): ArrayList<*> {
        val result = ArrayList<FoundedSequence>()
        var entry: Node?
        var sequence: FoundedSequence
        var counter = 0
        val entriesLength = entries.length
        var handledEntryCounter = 0
        while (entries.item(counter).also { entry = it } != null) {
            ++handledEntryCounter
            sequence = FoundedSequence()
            sequence.title =
                (xPath.evaluate("./title", entry, XPathConstants.NODE) as Node).textContent
            sequence.content =
                (xPath.evaluate("./content", entry, XPathConstants.NODE) as Node).textContent
            sequence.link = (xPath.evaluate(
                "./link",
                entry,
                XPathConstants.NODE
            ) as Node).attributes.getNamedItem("href").textContent
            result.add(sequence)
            counter++
        }
        return result
    }
}